### PR TITLE
Query: Adds EnableOptimisticDirectExecution flag to QueryRequestOptions

### PR DESCRIPTION
**Pull request by akotalwar**
# Pull Request Template

## Description

Adds new flag to QueryRequestOptions to allow customers to use the OptimisticDirectExecution pipeline.

## Type of change

Please delete options that are not relevant.

- [] New feature (non-breaking change which adds functionality)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs [hunk 1]
```diff
@@ -752,7 +753,7 @@ private static Documents.PartitionKeyDefinition GetPartitionKeyDefinition(InputP
             ContainerQueryProperties containerQueryProperties,
             ITrace trace)
         {
-            if (inputParameters.TestInjections == null || !inputParameters.TestInjections.EnableOptimisticDirectExecution) return null;
+            if (!inputParameters.EnabledOptimisticDirectExecution) return null;
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs:756]: >d [](http://example.com/codeflow?start=39&length=1)

Please remove.

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 2]
```diff
@@ -43,6 +43,14 @@ public class QueryRequestOptions : RequestOptions
         /// </value>
         public bool? EnableLowPrecisionOrderBy { get; set; }
 
+        /// <summary>
+        /// Gets or sets the option for customers to opt in or out of the new OptimisticDirectExecution pipeline.
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs:47]: >new [](http://example.com/codeflow?start=74&length=3)

Please remove.

Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs:47]: >the [](http://example.com/codeflow?start=70&length=3)

Nit, not needed.

Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs:47]: >Gets or sets the option for customers to opt in or out of the new OptimisticDirectExecution pipeline [](http://example.com/codeflow?start=12&length=100)

Gets or sets the option for customers to opt in for direct (optimistic) execution of the query.

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 3]
```diff
@@ -43,6 +43,14 @@ public class QueryRequestOptions : RequestOptions
...
+        /// </summary>
+        /// <value>
+        /// The OptimisticDirectExecution pipeline offers improved performance for single partition queries when compared to existing pipelines.
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs:50]: Direct (optimistic) execution offers improved performance for several kinds of queries such as a single partition streaming query.

**Review by adityasa — APPROVED** [review]
:shipit:

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs [hunk 4]
```diff
@@ -824,6 +825,7 @@ public sealed class InputParameters
                 ExecutionEnvironment? executionEnvironment,
                 bool? returnResultsInDeterministicOrder,
                 bool forcePassthrough,
+                bool enabledOptimisticDirectExecution,
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs:828]: >enabled [](http://example.com/codeflow?start=21&length=7)

Please change from `enabled` to `enable`

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs [hunk 5]
```diff
@@ -873,6 +876,7 @@ public sealed class InputParameters
             public bool ReturnResultsInDeterministicOrder { get; }
             public TestInjections TestInjections { get; }
             public bool ForcePassthrough { get; }
+            public bool EnabledOptimisticDirectExecution { get; }
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs:879]: >Enabled [](http://example.com/codeflow?start=24&length=7)

please change from `Enabled` to `Enable`

**Review by neildsh — DISMISSED** [review]

**Review by sc978345 — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs [hunk 6]
```diff
@@ -43,6 +43,14 @@ public class QueryRequestOptions : RequestOptions
         /// </value>
         public bool? EnableLowPrecisionOrderBy { get; set; }
 
+        /// <summary>
+        /// Gets or sets the option for customers to opt in for direct (optimistic) execution of the query..
+        /// </summary>
+        /// <value>
+        /// Direct (optimistic) execution offers improved performance for several kinds of queries such as a single partition streaming query.
+        /// </value>
+        internal bool EnableOptimisticDirectExecution { get; set; } = false;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/RequestOptions/QueryRequestOptions.cs:52]: >false [](http://example.com/codeflow?start=70&length=5)

please remove

**Review by sboshra — DISMISSED** [review]
:shipit:

**Review by sboshra — APPROVED** [review]

**Pull request merged**